File: go/editor.go
package cjkfixtures

import "unicode/utf8"

// Editor is a minimal, framework-agnostic model of an IME-aware text input,
// mirroring the DOM lifecycle a real input goes through during CJK composition:
// CompositionStart -> CompositionUpdate(s) -> CompositionEnd, plus Keydown,
// Blur and Focus. The bugs map lets a fixture build a deliberately broken
// editor and prove the test catches it.
type Editor struct {
	Value         string
	Composing     bool
	ComposingData string
	Submitted     int
	bugs          map[string]bool
}

// NewEditor builds an editor. Pass bug names ("enterDuringCompose",
// "dropOnBlur", "stuckComposing") to construct a broken one; none = correct.
func NewEditor(bugs ...string) *Editor {
	m := make(map[string]bool, len(bugs))
	for _, b := range bugs {
		m[b] = true
	}
	return &Editor{bugs: m}
}

// Keydown handles a key press. isComposing mirrors KeyboardEvent.isComposing:
// true while the IME owns the keystroke (e.g. Enter confirming a candidate).
func (e *Editor) Keydown(key string, isComposing bool) {
	midComposition := e.Composing || isComposing

	if key == "Enter" {
		// Failure mode #1: the Enter that confirms an IME candidate must not
		// also submit.
		if e.bugs["enterDuringCompose"] {
			e.Submitted++ // buggy: ignores isComposing
			return
		}
		if !midComposition {
			e.Submitted++
		}
		return
	}

	// While composing, text arrives via composition events, not raw keydown.
	if !e.Composing && utf8.RuneCountInString(key) == 1 {
		e.Value += key
	}
}

func (e *Editor) CompositionStart() {
	e.Composing = true
	e.ComposingData = ""
}

func (e *Editor) CompositionUpdate(data string) {
	e.ComposingData = data
}

func (e *Editor) CompositionEnd(data string) {
	e.Value += data
	e.Composing = false
	e.ComposingData = ""
}

func (e *Editor) Blur() {
	// Failure mode #5: leaving composition state stuck so later input is lost.
	if e.bugs["stuckComposing"] {
		return // buggy: never resets, CompositionEnd never arrives
	}
	// Failure mode #4: dropping the pending composition instead of committing.
	if e.bugs["dropOnBlur"] {
		e.Composing = false
		e.ComposingData = "" // buggy: silently drops pending text
		return
	}
	// Correct: finalize the pending composition so nothing is lost.
	if e.Composing {
		e.Value += e.ComposingData
		e.Composing = false
		e.ComposingData = ""
	}
}

func (e *Editor) Focus() {
	if e.bugs["stuckComposing"] {
		return // does not clear stuck state
	}
	e.Composing = false // correct: start from a clean composition state
}

File: js/src/editor.js
// A minimal, framework-agnostic model of an IME-aware text input.
//
// It reproduces the lifecycle a real DOM input goes through during composition
// for any IME — Japanese kana, Korean Hangul, Chinese pinyin candidates:
// compositionstart -> compositionupdate(s) -> compositionend, plus keydown /
// blur / focus. The `bugs` set lets a fixture construct a deliberately broken
// editor and prove the test catches it.
//
// Adapt the event methods to your real component (wire `keydown` to your
// keydown handler, `compositionStart/Update/End` to the matching DOM events,
// etc.) and the assertions in /test carry straight over. See README for a DOM
// adapter using real CompositionEvent objects.

export function createEditor(opts = {}) {
  return new ImeInput(new Set(opts.bugs || []));
}

class ImeInput {
  constructor(bugs) {
    this.bugs = bugs;
    this.value = "";
    this.composing = false;
    this.composingData = "";
    this.submitted = 0;
  }

  // A key was pressed. `isComposing` mirrors KeyboardEvent.isComposing — it is
  // true while the IME owns the keystroke (e.g. Enter confirming a candidate).
  keydown(key, isComposing = false) {
    const midComposition = this.composing || isComposing;

    if (key === "Enter") {
      // Failure mode #1 (early-Enter compose-confirm): the Enter that confirms
      // an IME candidate must NOT also submit / insert a newline.
      if (this.bugs.has("enterDuringCompose")) {
        this.submitted++; // buggy: ignores isComposing
        return;
      }
      if (!midComposition) this.submitted++;
      return;
    }

    // While composing, the IME produces text via composition events, not via
    // raw keydown, so a printable keydown here is part of candidate selection.
    if (!this.composing && key.length === 1) {
      this.value += key;
    }
  }

  compositionStart() {
    this.composing = true;
    this.composingData = "";
  }

  compositionUpdate(data) {
    this.composingData = data;
  }

  compositionEnd(data) {
    this.value += data;
    this.composing = false;
    this.composingData = "";
  }

  blur() {
    // Failure mode #5 (re-entry after blur): some browsers never fire
    // compositionend when focus leaves mid-composition, leaving `composing`
    // stuck true so all later input is ignored.
    if (this.bugs.has("stuckComposing")) {
      return; // buggy: never resets, compositionend never arrives
    }
    // Failure mode #4 (commit-callback-drop on focus-shift): the pending
    // composition is thrown away instead of committed.
    if (this.bugs.has("dropOnBlur")) {
      this.composing = false;
      this.composingData = ""; // buggy: silently drops pending text
      return;
    }
    // Correct: finalize the pending composition so nothing is lost.
    if (this.composing) {
      this.value += this.composingData;
      this.composing = false;
      this.composingData = "";
    }
  }

  focus() {
    if (this.bugs.has("stuckComposing")) return; // does not clear stuck state
    this.composing = false; // correct: start from a clean composition state
  }
}
